rewrite-core/src/main/java/org/openrewrite/internal/JavaTypeReferences.java

```diff
@@ -0,0 +1,45 @@
+/*
+ * Copyright 2024 the original author or authors.
+ * <p>
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * <p>
+ * https://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.openrewrite.internal;
+
+import lombok.AccessLevel;
+import lombok.Getter;
+import lombok.RequiredArgsConstructor;
+import org.openrewrite.JavaTypeReferenceProvider;
+import org.openrewrite.SourceFile;
+import org.openrewrite.trait.JavaTypeReference;
+
+import java.util.HashSet;
+import java.util.ServiceLoader;
+import java.util.Set;
+
+@RequiredArgsConstructor(access = AccessLevel.PRIVATE)
+@Getter
+public class JavaTypeReferences {
+    private final SourceFile document;
+    private final Set<JavaTypeReference> typeReferences;
+
+    public static JavaTypeReferences build(SourceFile doc) {
+        Set<JavaTypeReference> typeReferences = new HashSet<>();
+        ServiceLoader<JavaTypeReferenceProvider> loader = ServiceLoader.load(JavaTypeReferenceProvider.class);
```

Comment:
We might have to use a different classloader here, which is possible because this method has an overloaded variant that takes a classloader as argument